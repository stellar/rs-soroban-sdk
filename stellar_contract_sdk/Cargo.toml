[package]
name = "stellar_contract_sdk"
version = "0.0.1"
authors = ["Stellar Development Foundation <info@stellar.org>"]
license = "Apache-2.0"
edition = "2021"

[lib]
crate-type = ["cdylib", "rlib"]
