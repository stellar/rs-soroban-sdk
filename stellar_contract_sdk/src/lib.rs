#![cfg_attr(target_family = "wasm", no_std)]
#![allow(dead_code)]

// This requires running on unstable.
//#![feature(alloc)]
//mod alloc;

mod bignum;
mod bitset;

#[cfg(target_family = "wasm")]
mod host;

#[cfg(not(target_family = "wasm"))]
pub mod testing;
#[cfg(not(target_family = "wasm"))]
use testing::host;

pub mod ledger;
mod map;
mod object;
mod or_abort;
mod result;
mod rt;
mod status;
mod vec;

mod symbol;
mod val;

pub use bignum::BigNum;
pub use bitset::BitSet;
pub use map::Map;
use object::ObjType;
pub use object::Object;
pub use or_abort::OrAbort;
pub use result::OpResult;
pub use status::Status;
pub use symbol::Symbol;
pub use val::Val;
pub use vec::Vec;

#[inline(always)]
pub fn require(b: bool) {
    b.or_abort();
}

#[inline(always)]
pub fn log_value(v: Val) {
    unsafe {
        host::context::log_value(v);
    }
}

#[inline(always)]
pub fn call0(contract: Val, func: Symbol) -> Val {
    unsafe { host::call::call0(contract, func.into()) }
}

#[inline(always)]
pub fn call1(contract: Val, func: Symbol, a: Val) -> Val {
    unsafe { host::call::call1(contract, func.into(), a) }
}

#[inline(always)]
pub fn call2(contract: Val, func: Symbol, a: Val, b: Val) -> Val {
    unsafe { host::call::call2(contract, func.into(), a, b) }
}

#[inline(always)]
pub fn call3(contract: Val, func: Symbol, a: Val, b: Val, c: Val) -> Val {
    unsafe { host::call::call3(contract, func.into(), a, b, c) }
}

#[inline(always)]
pub fn call4(contract: Val, func: Symbol, a: Val, b: Val, c: Val, d: Val) -> Val {
    unsafe { host::call::call4(contract, func.into(), a, b, c, d) }
}

#[inline(always)]
pub fn get_last_operation_result() -> OpResult {
    unsafe { OpResult::unchecked_from_obj(host::context::get_last_operation_result()) }
}
