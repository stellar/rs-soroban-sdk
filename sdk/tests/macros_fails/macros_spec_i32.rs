use stellar_contract_sdk::contractimpl;

pub struct MyType<A>(pub A);

pub struct Contract;

#[contractimpl]
impl Contract {
    pub fn add(a: MyType<i32>, b: MyType<i32>) -> i8 {
        unimplemented!()
    }
}

pub fn main() {}
