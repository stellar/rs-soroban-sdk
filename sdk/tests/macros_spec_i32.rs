#![cfg(feature = "testutils")]

use std::io::Cursor;

use stellar_contract_sdk::{contractimpl, Env, IntoVal, TryFromVal};
use stellar_xdr::{ReadXdr, ScSpecEntry, ScSpecFunctionV0, ScSpecTypeDef};

pub struct Contract;

#[contractimpl]
impl Contract {
    pub fn add(a: i32, b: i32) -> i32 {
        a + b
    }
}

#[test]
fn test_functional() {
    let e = Env::default();
    let a = 10i32.into_val(&e);
    let b = 12i32.into_val(&e);
    let c = __add::call_raw(e.clone(), a, b);
    let c = i32::try_from_val(&e, c).unwrap();
    assert_eq!(c, 22);
}

#[test]
fn test_spec() {
    let entries = ScSpecEntry::read_xdr(&mut Cursor::new(&__SPEC_XDR_ADD)).unwrap();
    let expect = ScSpecEntry::FunctionV0(ScSpecFunctionV0 {
        name: "add".try_into().unwrap(),
        input_types: vec![ScSpecTypeDef::I32, ScSpecTypeDef::I32]
            .try_into()
            .unwrap(),
        output_types: vec![ScSpecTypeDef::I32].try_into().unwrap(),
    });
    assert_eq!(entries, expect);
}
