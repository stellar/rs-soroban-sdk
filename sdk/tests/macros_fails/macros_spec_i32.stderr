error: generics unsupported on user-defined types in contract functions
 --> tests/macros_fails/macros_spec_i32.rs:6:21
  |
6 | pub fn add(a: MyType<i32>, b: MyType<i32>) -> i8 {
  |                     ^

error: generics unsupported on user-defined types in contract functions
 --> tests/macros_fails/macros_spec_i32.rs:6:37
  |
6 | pub fn add(a: MyType<i32>, b: MyType<i32>) -> i8 {
  |                                     ^
